File: sentil-java/src/main/java/io/github/sedislab/sentil/SimModel.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;

/** A declarative stochastic model, one init and one advance expression per variable. */
public final class SimModel extends NativeResource {
    SimModel(long handle) {
        super(handle, NativeLib::simModelDestroy);
    }

    /**
     * Build a model over the named variables stepping by dt for horizon steps. The
     * init, advance, and noise arrays are consumed.
     */
    public static SimModel create(String[] variables, double dt, long horizon, SimExpr[] init,
            SimExpr[] advance, NoiseModel[] noise) throws SentilException {
        long[] initHandles = collect(init);
        long[] advanceHandles = collect(advance);
        long[] noiseHandles = collect(noise);
        disownAll(init);
        disownAll(advance);
        disownAll(noise);
        return new SimModel(NativeLib.simModelCreate(variables, dt, horizon, initHandles,
                advanceHandles, noiseHandles));
    }

    private static long[] collect(NativeResource[] resources) {
        long[] handles = new long[resources.length];
        for (int i = 0; i < resources.length; i++) {
            handles[i] = resources[i].handle();
        }
        return handles;
    }

    private static void disownAll(NativeResource[] resources) {
        for (NativeResource resource : resources) {
            resource.disown();
        }
    }

    /** Simulate one full-horizon trajectory from a seed. */
    public Trace simulate(long seed) throws SentilException {
        return new Trace(NativeLib.simModelSimulate(handle(), seed));
    }

    /** The state variable names. */
    public List<String> variables() throws SentilException {
        return Arrays.asList(NativeLib.simModelVariables(handle()));
    }

    /** The time step. */
    public double dt() {
        return NativeLib.simModelDt(handle());
    }

    /** The number of steps in a trajectory. */
    public long horizon() {
        return NativeLib.simModelHorizon(handle());
    }

    /** Convert to a stochastic system for the CPU rare-event path. */
    public StochasticSystem toStochasticSystem() throws SentilException {
        return new StochasticSystem(NativeLib.simModelToStochasticSystem(handle()));
    }
}

File: sentil-matlab/+sentil/SimModel.m
classdef SimModel < handle
    %SIMMODEL A declarative stochastic model with one init and advance rule per variable.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods (Static)
        function m = create(variables, dt, horizon, init, advance, noise)
            %CREATE A model over the named variables. The expressions and noise models are consumed.
            if ~iscell(variables)
                variables = cellstr(variables);
            end
            m = sentil.SimModel(sentil_mex('sim_model_create', variables, double(dt), ...
                double(horizon), consume_handles(init), consume_handles(advance), ...
                consume_handles(noise)));
        end
    end

    methods
        function obj = SimModel(handle)
            obj.Handle = handle;
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('sim_model_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function t = simulate(obj, seed)
            %SIMULATE One full-horizon trajectory from a seed.
            obj.assertOpen();
            t = sentil.Trace(uint64(sentil_mex('sim_model_simulate', obj.Handle, double(seed))));
        end

        function v = variables(obj)
            %VARIABLES The model's variable names.
            obj.assertOpen();
            v = sentil_mex('sim_model_variables', obj.Handle);
        end

        function d = dt(obj)
            %DT The time step.
            obj.assertOpen();
            d = sentil_mex('sim_model_dt', obj.Handle);
        end

        function h = horizon(obj)
            %HORIZON The trajectory length, in steps.
            obj.assertOpen();
            h = sentil_mex('sim_model_horizon', obj.Handle);
        end

        function s = to_stochastic_system(obj)
            %TO_STOCHASTIC_SYSTEM A sampling-ready system for rare-event estimation.
            obj.assertOpen();
            s = sentil.StochasticSystem(uint64(sentil_mex('sim_model_to_stochastic_system', ...
                obj.Handle)));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this model has been closed');
            end
        end
    end
end